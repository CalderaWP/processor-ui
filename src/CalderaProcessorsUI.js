import React from 'react';
import ReactDOM from 'react-dom';
import CalderaProcessorsWithState from './CalderaProcessorsWithState';
import { Provider } from 'react-redux';
import {processorsStore} from './state/processorsStore';
import {validationSubscriber} from "./state/validationSubscriber";

/**
 * Manages a single instance of the Caldera Processor UI with state and validation.
 *
 * @param {String} elementId ID of element UI gets mounted on.
 *
 * @constructor
 */
export default function CalderaProcessorsUI(elementId){
	let validationSubscriptionUnsubscribe = null;

	/**
	 * Get the processor store
	 *
	 * @return {Object}
	 */
	this.getProcessorStore =() => {
		return processorsStore;
	};

	/**
	 * Add subscriber to store used for validation
	 * @return {boolean}
	 */
	this.subscribeForValidation = () => {
		validationSubscriptionUnsubscribe = processorsStore.subscribe(validationSubscriber);
		return 'function' === typeof validationSubscriptionUnsubscribe
			? true
			: false;

	};

	/**
	 * Unsubscribe the validation subscriber
	 *
	 * @return {boolean}
	 */
	this.unSubscribeToProcessorValidation = () => {
		if( 'function' === typeof  validationSubscriptionUnsubscribe ){
			validationSubscriptionUnsubscribe();
			validationSubscriptionUnsubscribe = null;
			return true;
		}
		return false;
	};

	/**
	 * Creates the processors componentWithState with state setup properly
	 * @return {*}
	 */
	this.componentWithState = () => {
		return(
			<Provider store={processorsStore}>
				<CalderaProcessorsWithState />
			</Provider>
		);
	};

	/**
	 * Mounts the component, with state, on the element whose ID was passed in constructor
	 */
	this.mountOnDOM = () => {
		this.subscribeForValidation();
		ReactDOM.render(
			this.componentWithState(),
			document.getElementById(elementId)
		);
	};

}


