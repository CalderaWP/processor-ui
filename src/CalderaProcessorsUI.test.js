import React from 'react';
import ReactDOM from 'react-dom';
import CalderaProcessorsUI from './CalderaProcessorsUI';
import renderer from 'react-test-renderer';

describe( 'Processor UI factory', () => {
	it( 'Mounts without crashing', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-1');
		ReactDOM.render(
			CalderaProcessors.componentWithState(),
			document.createElement('processor-ui-test-1')
		);

	});

	it( 'Renders without crashing', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-1b');
		const component = renderer.create(
			<div>
				{CalderaProcessors.componentWithState()}
			</div>
		);
		expect( component.toJSON() ).toMatchSnapshot();
	});

	it( 'Can be used to access processor store', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-2');
		expect( typeof CalderaProcessors.getProcessorStore() ).toEqual('object' );
	});

	it( 'Could be used to subscribe to processor store', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-3');
		expect( typeof CalderaProcessors.getProcessorStore().subscribe ).toEqual('function' );
	});

	it( 'Could be used to dispatch to processor store', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-4');
		expect( typeof CalderaProcessors.getProcessorStore().dispatch ).toEqual('function' );
	});

	it( 'Can setup validation subscription', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-5');
		expect( CalderaProcessors.subscribeForValidation() ).toEqual(true);
	});

	it( 'Can setup validation subscription and then unsubscribe', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-6');
		CalderaProcessors.subscribeForValidation();
		expect( CalderaProcessors.unSubscribeToProcessorValidation() ).toEqual(true);
	});

	it( 'returns false instead of making errors if unsubscribe is called before subscribe', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-7');
		expect( CalderaProcessors.unSubscribeToProcessorValidation() ).toEqual(false);
	});

	it( 'Stops thinking its subscribed when its no longer subscribed', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-8');
		CalderaProcessors.subscribeForValidation();
		CalderaProcessors.unSubscribeToProcessorValidation();
		expect( CalderaProcessors.unSubscribeToProcessorValidation() ).toEqual(false);
	});

	it( 'Returns a object for the processor store', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-9');
		expect( typeof CalderaProcessors.getProcessorStore() ).toEqual('object');
	});

	it( 'Renders without crashing', () => {
		const CalderaProcessors = new CalderaProcessorsUI('processor-ui-test-1b');
		const component = renderer.create(
			<div>
				{CalderaProcessors.componentWithState()}
			</div>
		);
		expect( component.toJSON() ).toMatchSnapshot();
	});
});