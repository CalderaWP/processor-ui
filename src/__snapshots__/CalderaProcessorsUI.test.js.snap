// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processor UI factory Renders without crashing 1`] = `
<div>
  <div
    className="caldera-forms-processors"
  >
    <div />
    <div>
      <button
        onClick={[Function]}
      >
        NewProcessor Processor
      </button>
    </div>
  </div>
</div>
`;

exports[`Processor UI factory Renders without crashing 2`] = `
<div>
  <div
    className="caldera-forms-processors"
  >
    <div />
    <div>
      <button
        onClick={[Function]}
      >
        NewProcessor Processor
      </button>
    </div>
  </div>
</div>
`;
